### PR TITLE
Add feature to skip unavailable fragments

### LINKED ISSUE #10448: [hlsnative] Add an ability to skip unavailable fragments
- [x] Bug report (encountered problems with youtube-dl)

---

```
PS C:\dev\youtube-dl\master> youtube-dl.exe https://www.twitch.tv/naysayer88/v/85713845 -
v
[debug] System config: []
[debug] User config: []
[debug] Command-line args: ['https://www.twitch.tv/naysayer88/v/85713845', '-v']
[debug] Encodings: locale cp1251, fs mbcs, out cp866, pref cp1251
[debug] youtube-dl version 2016.08.24.1
[debug] Python version 3.4.4 - Windows-10-10.0.10240
[debug] exe versions: ffmpeg N-81192-g04da20e, ffprobe N-81192-g04da20e, rtmpdump 2.4
[debug] Proxy map: {}
[twitch:vod] 85713845: Downloading vod info JSON
[twitch:vod] 85713845: Downloading vod access token
[twitch:vod] 85713845: Downloading m3u8 information
[debug] Invoking downloader on 'http://vod.edgecast.hls.ttvnw.net/v1/AUTH_system/vods_60f4/naysayer88_23003365072_506378255/chunked/index-dvr.m3u8'
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 43
[download] Destination: Gammin' - Obduction Spoilerama-v85713845.mp4
[download]  34.9% of ~13.53MiB at Unknown speed ETA Unknown ETAERROR: unable to download video data: HTTP Error 400: Bad Request
Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpvarrvgfb\build\youtube_dl\YoutubeDL.py", line 1694, in process_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpvarrvgfb\build\youtube_dl\YoutubeDL.py", line 1636, in dl
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpvarrvgfb\build\youtube_dl\downloader\common.py", line 354, in download
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpvarrvgfb\build\youtube_dl\downloader\hls.py", line 102, in real_download
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpvarrvgfb\build\youtube_dl\downloader\common.py", line 354, in download
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpvarrvgfb\build\youtube_dl\downloader\http.py", line 58, in real_download
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpvarrvgfb\build\youtube_dl\YoutubeDL.py", line 1996, in urlopen
  File "C:\Python\Python34\lib\urllib\request.py", line 470, in open
    response = meth(req, response)
  File "C:\Python\Python34\lib\urllib\request.py", line 580, in http_response
    'http', request, response, code, msg, hdrs)
  File "C:\Python\Python34\lib\urllib\request.py", line 508, in error
    return self._call_chain(*args)
  File "C:\Python\Python34\lib\urllib\request.py", line 442, in _call_chain
    result = func(*args)
  File "C:\Python\Python34\lib\urllib\request.py", line 588, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 400: Bad Request
```

---

Twitch may serve unavailable fragments causing download to fail while using hlsnative.
ffmpeg skips unavailable fragments by default. Same feature should be available in youtube-dl.

### PR DESCRIPTION
**Pull request by dstftw**
### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] New feature

---

**tl;dr:**
This PR introduces a new option `--skip-unavailable-fragments` that allows to skip fragments that are not available.

---

Fragment based delivery protocols may serve fragments that are not available. For example, twitch may serve constantly unavailable fragments (see https://github.com/rg3/youtube-dl/issues/10448, constant HTTP 400 error), another example is livestream.com (see https://github.com/rg3/youtube-dl/issues/10165, temporary HTTP 500 error). Also youtube sometimes serves temporary HTTP 404 error that is ususally fixed with subsequent retry.

Currently a fragment is considered unavailable if all `--fragment-retries` download attempts have failed with HTTP error.
Currently only dashsegments and hlsnative downloaders respect `--fragment-retries` and `--skip-unavailable-fragments`.

Fixes #10448 and #10165.

### CODE REVIEW

**Comment by remitamine** [comment]
i think it will be better to create a method(for example `_download_fragment`) to share the code and enable it also for f4m downloade in case that this kind of errors happen also for f4m fragments.
i will try to do the same for #9937 as it can be useful for other fragment downloaders.

**Comment by dstftw** [comment]
This will require major refactoring of all fragment downloaders. The intention of this PR is to fix actual bugs in the first place.

**Comment by remitamine** [comment]
> This will require major refactoring of all fragment downloaders. The intention of this PR is to fix actual bugs in the first place.

i don't think that it will require a lot of changes to make it work for all of the extractors, i already made the changes in #9937 by adding two methods(`_download_fragment` and `_append_fragment`) to `FragmentFD` and calling them from the other fragment downloaders, and the code for fragment skipping and retries can be moved to `_download_fragment` and it should work without a problem.

**Comment by dstftw** [comment]
With very brief look it will require at least:
1. Generalizing fragments representation and handling across all extractors (currently some extractors use fragment list some parse manifest on fly).
2. Generalizing live stream support (since f4m supports it) and implementing live support in other downloaders to make sure this generalization works correctly.
3. Considering how `--fragment-retry` and `--skip-unavailable-fragments` should work for live streams.
These changes will require more testing and will affect all fragment downloaders introducing more probability for breakages and regressions.

**Comment by remitamine** [comment]
> These changes will require more testing and will affect all fragment downloaders introducing more probability for breakages and regressions.

i don't thnk that this is required to be done in one step,whenever there is a part that all the extractors support it can be moved to `FragmentFD`.
as you mention the case of live stream it may not be ready to be generalized yet, for the change in #9937 live streams should be a problem.

**Comment by dstftw** [comment]
>i don't thnk that this is required to be done in one step

This is why this PR is opened. Generic problems should be tackled step by step. The intention of it is to introduce missing options and fix vital problems. The next step can be improving and refactoring.

**Comment by yan12125** [comment]
LGTM. Could you rebase and push it to rg3/master?

**Comment by dstftw** [comment]
Done.

**Comment by yan12125** [comment]
Thanks!
